polkadot/node/core/backing/src/lib.rs

```diff
@@ -1105,20 +1232,31 @@ async fn construct_per_relay_parent_state<Context>(
 
 	let session_index = try_runtime_api!(session_index);
 
-	let inject_core_index = request_node_features(parent, session_index, ctx.sender())
+	let validators_arc = per_session_cache
+		.get_or_fetch_validators(session_index, parent, ctx.sender())
+		.await?
+		.unwrap();
+	let validators = validators_arc.as_ref().clone();
+
+	let node_features_arc = per_session_cache
+		.get_or_fetch_node_features(session_index, parent, ctx.sender())
 		.await?
-		.unwrap_or(NodeFeatures::EMPTY)
+		.unwrap();
+
+	let inject_core_index = node_features_arc
```

Comment:
node features are cheap since they're expressed as a bitvec, so we're probably fine without and `Arc` here

Reply:
Okay, removed `Arc`.